maturity-matrix.yaml

```diff
@@ -0,0 +1,161 @@
+# OpenTelemetry is a complex project with many moving parts. Thankfully, we
+# strive to keep those parts loosely coupled where we can (this has been a goal
+# since the very announcement of OpenTelemetry, per this blog post:
+# https://medium.com/opentracing/merging-opentracing-and-opencensus-f0fe9c7ca6f0)
+#
+# This file describes the structure of the project and its sub-components,
+# definitions of API and stability maturity levels, and finally records the
+# current maturity of each component. We maintain this file as YAML in order to
+# make it easier to automate scripts (e.g., for the website) that present the
+# data to OpenTelemetry end-users.
+
+
+###############################################################################
+# MATURITY LEVELS
+#
+# OpenTelemetry components advertise their own maturity along two axes:
+# - PRODUCTION MATURITY: How reliable is this component from a
+#   production-readiness standpoint?
+# - API MATURITY: How stable is the component's API from a backwards- and
+#   forwards-compatibility standpoint)?
+#
+# In some cases, for instance, OpenTelemetry components may be immature from an
+# API maturity standpoint, yet be nearly 100% safe from a production standpoint
+# – and vice versa. Similarly, end-users may have more tolerance for lost
+# development time or software stability depending on their situation, so we
+# strive to separate self-reported maturity along these two axes.
+#
+# Needless to say, representations here are best-effort, and there is no
+# substitute for a mature release process. I.e., don't blindly deploy "mature"
+# OpenTelemetry components (or any software, for that matter!).
+productionMaturityLevels:
+  # "unknown" maturity is just what it purports to be. It's reasonable to
+  # assume "unstable", but we separate "unknown" and "unstable" to distinguish
+  # between a lack of evidence about production maturity and actual evidence of
+  # immaturity.
+  - unknown
+  # "unstable" components are not recommended for production workloads. They
+  # may crash the process, introduce performance artifacts, or have known
+  # security issues.
+  - unstable
+  # "beta" components have been used successfully in production workloads at
+  # scale, but either due to the uncertainty introduced by active development
+  # or a small number of production deployments, they should still be used
+  # cautiously in new deployments. They do not have known security issues.
+  - beta
+  # "stable" components are, to the best of the author's knowledge, safe for
+  # typical production use cases.
+  - stable
+
+# All OpenTelemetry APIs follow semver conventions (i.e., after v1.x,
+# backwards-incompatible changes should bump the major version number).
+apiMaturityLevels:
+  # "unimplemented" APIs do not exist yet for the component/language.
+  - unimplemented
```

Comment:
should there also be "N/A". Do we expect auto-instrumentation for the GoLang for instance?

Reply:
Yes, auto-instrumentation for golang includes wrappers for gorillamux, sqlx, etc. -- so it's not _fully_ automatic but it's O(1) lines of code to add.

Reply:
Yeah. But than it falls under "integrations" bucket and we may want more than overall status? Perhaps those integrations can also contribute to the registry that @austinlparker is working on?

Reply:
Yeah, I think auto-instrumentation is just a type of integration in the registry.

Reply:
@lizthegrey re

>auto-instrumentation for golang includes wrappers for gorillamux, sqlx, etc. -- so it's not fully automatic but it's O(1) lines of code to add.

I'm sympathetic to the fact that it's hard to do better than this ^^ in Go, but I (unfortunately) don't think 1-line changes "count" for brownfield auto-instrumentation in many environments. This would support some sort of "n/a" designation per @SergeyKanzhelev's comment.

An alternative would be to define different levels of auto-instrumentation: perhaps "unsupported", "smallCode", and "noCode". Or, alternatively, to define auto-instrumentation criteria differently for different languages (with their different expectations).

Reply:
Either n/a or just move it to the registry